File: 프로그래머스/lv3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    
    dx = (0, 0, -1, 1)
    dy = (-1, 1, 0, 0)
    
    # 좌표 개념으로 이동하기위해 크기를 두배 늘려줌
    characterX *= 2; characterY *= 2; itemX *= 2; itemY *= 2
    board = [[False] * (51*2) for _ in range(51*2)]
    
    for r in rectangle:
        # 직사각형의 테두리 및 내부 모두 True로
        for x in range(r[0]*2, r[2]*2+1):
            for y in range(r[1]*2, r[3]*2+1):
                board[x][y] = True
                
    for r in rectangle:
        # 직사각형의 테두리만 남기고 모두 False로
        for x in range(r[0]*2+1, r[2]*2):
            for y in range(r[1]*2+1, r[3]*2):
                board[x][y] = False
    
    queue = deque([]); queue.append((characterX, characterY, 0))
    visit = [[False] * (51*2) for _ in range(51*2)]; visit[characterX][characterY] = True
    
    while queue:
        x, y, cnt = queue.popleft()
        
        if (x, y) == (itemX, itemY):
            return cnt/2
        
        for i in range(4):
            nx = x + dx[i]; ny = y + dy[i]
            if 0 <= nx < (51*2) and 0 <= ny < (51*2):
                if not visit[nx][ny] and board[nx][ny]:
                    visit[nx][ny] = True
                    queue.append((nx, ny, cnt+1))
            

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.cpp
#include <string>
#include <vector>
#include <deque>
#include <tuple>

using namespace std;

int solution(vector<vector<int>> rectangle, int characterX, int characterY, int itemX, int itemY) {
    
    int dx[4] = {0, 0, -1, 1};
    int dy[4] = {-1, 1, 0, 0};
    
    characterX *= 2; characterY *= 2; itemX *= 2; itemY *= 2;
    vector<vector<bool>> board(51*2, vector<bool>(51*2, false));
    
    for (auto rec: rectangle)
    {
        for (int x = rec[0]*2; x <= rec[2]*2; ++x)
        {
            for (int y = rec[1]*2; y <= rec[3]*2; ++y)
            {
                board[x][y] = true;
            }
        }
    }
    
    for (auto rec: rectangle)
    {
        for (int x = rec[0]*2+1; x < rec[2]*2; ++x)
        {
            for (int y = rec[1]*2+1; y < rec[3]*2; ++y)
            {
                board[x][y] = false;
            }
        }
    }
    
    deque<tuple<int, int, int>> queue; queue.push_back({characterX, characterY, 0});
    vector<vector<bool>> visit(51*2, vector<bool>(51*2, false)); visit[characterX][characterY] = true;
    
    while (!queue.empty())
    {
        auto [x, y, cnt] = queue.front(); queue.pop_front();
        
        if (x == itemX && y == itemY)
        {
            return cnt / 2;
        }
        
        for (int i = 0; i < 4; ++i)
        {
            int nx = x + dx[i]; int ny = y + dy[i];
            if (0 <= nx && nx < 51*2 && 0 <= ny && ny < 51*2)
            {
                if (!visit[nx][ny] && board[nx][ny])
                {
                    visit[nx][ny] = true;
                    queue.push_back({nx, ny, cnt+1});
                }
            }
        }
    }
    
    return 0;
}